add integrity registers (#189)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

  Appraisal verifies that, for each value in `cryptokeys`, there is a matching Reference Value entry.
Matching is described in {{sec-cryptokeys-matching}}.

> * `integrity-registers` (index 14): A register with one or more banks containing measurements associated with the environment.  Described in {{sec-comid-integrity-registers}}.

Reviewer: ```suggestion
* `integrity-registers` (index 14):  Multiple registers with one or more banks containing measurements associated with the environment.  Described in {{sec-comid-integrity-registers}}.
```

{::include cddl/crypto-key-type-choice.cddl}
~~~

##### Integrity Registers {#sec-comid-integrity-registers}

> An Integrity register is a group of one or more named measurements.

Reviewer: ```suggestion
Integrity registers is a group of one or more named measurements.
```

Author: reworded in [0daa574](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/189/commits/0daa57453d485e1ad42dd2d338868d0b3d389fcd)

> Each measurement has an unique identifier (either an unsigned integer or a string) and one or more associated digests.

Author: reworded in [0daa574](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/189/commits/0daa57453d485e1ad42dd2d338868d0b3d389fcd)

> This type can be used to model the PCRs in a TPM or vTPM, in which case the identifier is the register index, as well as other vendor-specific "Integrity Registers".

Reviewer: ```suggestion
Integrity registers can be used to model the PCRs in a TPM or vTPM ( in which case the identifier is the register index), or vendor-specific "Integrity Registers".
```

Author: reworded in [0daa574](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/189/commits/0daa57453d485e1ad42dd2d338868d0b3d389fcd)

> This type can be used to model the PCRs in a TPM or vTPM, in which case the identifier is the register index, as well as other vendor-specific "Integrity Registers".

Reviewer: It is not clear from line 117 in the change whether we are using `This type` to refer to the identifier or Integrity register specifically. Hence I tried to word smith it to add better clarity, however feel free to improve it further, if need be!

Reviewer 2: Somewhere in this section it might make sense to describe verifier behavior. I believe it is somewhat like ASN.1 SETS (vs. SEQUENCE) in that an entry "PCR5" at position 0 in the list should match a reference value even if the reference placed "PCR5" at position n>0. 

Maybe it goes without saying that verifier will not do type translation. That is if evidence has an entry `'5' : [...]` and reference values has an entry `5 : [...]` there is not ascii-to-numeric / numeric-to-ascii translation. What do you think?

Author: reworded in [0daa574](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/189/commits/0daa57453d485e1ad42dd2d338868d0b3d389fcd)

If all entries of the Reference Values array match a corresponding entry in the Accepted Claims Set array, then the `cryptokeys` Reference Value matches.
Otherwise, `cryptokeys` does not match.

##### Comparison for Integrity Registers {#sec-cmp-integrity-registers}

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer: ```suggestion
For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value. It is expected that the `type` of identifier in the Evidence is same as the one set in the Reference Values, as dictated by the profile.  Otherwise the comparison is undefined behaviour.
```

If all entries of the Reference Values array match a corresponding entry in the Accepted Claims Set array, then the `cryptokeys` Reference Value matches.
Otherwise, `cryptokeys` does not match.

> ##### Comparison for Integrity Registers {#sec-cmp-integrity-registers}

Reviewer 3: For other types of measurement, the reference value chooses which values to compare. Can this comparison work the same way?
Diffs below suggesting what the text would look like if this suggestion is adopted.

Author: > For other types of measurement, the reference value chooses which values to compare. Can this comparison work the same way?

I guess the crucial point is whether we want the integrity registers to support "quote" semantics a la TPM, i.e., we allow evidence to be a user-defined subset of the complete state of the attester's IRs.  If so, evidence dictates the direction of matching.  If not, i.e., if we can assume that the quote is always _complete_, then we can flip the direction.

Reviewer 2: The question is whether (a) the IR array should be considered an atomic measurement or (b) should each array member be atomic? If (b), then an IR as evidence could contain a superset of members contained in an RV array. The verifier should apply subset matching. If (a), then the matching semantic should be set equivalence. I don't feel strongly either way. If (a), then a second IR measurement can be used in evidence. However, if (b), the verifier would separate the IR into multiple IRs, and keep track of authority differences.

Author: Going back to our F2F discussion: *if* it can be assumed that a relying party cannot request a quote from the attester *independently* of the verifier - i.e., the RP cannot decide autonomously the geometry of the quote - then I am OK with @andrew-draper proposed text.

Reviewer: @nedmsmith : We would await a new version of suggested `text` from your side and then we can proceed here!

Reviewer 4: TL;DR the RP cannot decide autonomously the geometry of the quote and Andy's text is fine

I've been raising the question to the TCG IWG. There was a clear answer: the RP does not meddle with the geometry of the quote. The only usage scenario where that implicitly could be the case is where Attester owner and RP owner are collapsed roles and the local configuration of an Attester is required by certain RPs. In these cases the RP owner (which is also the Attester Owner) would also take on the role of Endorser and produce a CoRIM for the Verifier to take care of "the RP's needs". So, in the end the requirement would be again be addressed by the de-facto Endorser and not the RP.

> ##### Comparison for Integrity Registers {#sec-cmp-integrity-registers}

Reviewer: ```suggestion
##### Comparison for Integrity Registers {#sec-cmp-integrity-registers}
For each Integrity Register entry in the Reference Value, identified as: id => digests_{ref}, the Verifier will claim a "match" with Evidence only if the Evidence has an Integrity Register entry, identified as: id => digests_{ev} such that the `id` in Evidence and Reference Value both have the same type and value and the digests matches using the method defined in {{sec-cmp-digests}}.

If no entry is found, the Reference Value does not match.

Note that it is not required for all the Evidence entries in an Integrity Register array to be used during matching: Evidence could consist of a Target Environment's complete register space, while the Integrity Register array in Reference Values could consist of a subset of the Target Environment's register space.
```

Reviewer 2: > id => digests_{ev}

This nomenclature is confusing as it doesn't refer to the actual CDDL. Substitute `id` with `integrity-register-id-type-choice` and ` => digests_{ev}` with ` => [ + digest ]` and it would be more clear. Alternatively, include the aliased terms in the CDDL definition. E.g.:
```
integrity-registers = {
  + id: integrity-register-id-type-choice => digests: [ + digest ]
}
```
Then it would be OK to use `id` and `digests`.

Reviewer: > > the Verifier will claim a "match" with Evidence only if the Evidence has an Integrity Register entry
> 
> This wording is confusing because "Claim" has special meaning but it is used in the verb form. Other parts of the document use "match" in the verb form already. Maybe it's better to stay with this convention.
> 
> Nevertheless, the matching semantic is misrepresented by the sentence. The Verifier doesn't match evidence per se, as it accepts it simply because the (authenticated) Attester asserted it.
> 
> The Ref Val is accepted only if it is a subset of the Evidence.
> 
> The question that needs to be answered is whether or not the Ref Val subset requirements are applied to the IR array too. I think the answer should be yes.
> 
> The text I supplied seems to work in terms of describing applying the subset matching semantics to the IR array.

I agree we should use the term `match` only not mis-use the word claim.

Reviewer 2: +1

Author: > The question that needs to be answered is whether or not the Ref Val subset requirements are applied to the IR array too. I think the answer should be yes.

Given @henkbirkholz https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/189#discussion_r1475724199 I am OK too.

> The text I supplied seems to work in terms of describing applying the subset matching semantics to the IR array.

LGTM

Author: I suggest we accept Andy's proposed text and merge this.  As suggested by Yogesh, we can track the refinement put forward by Ned in a separate PR.

Reviewer: This works with me +1

Reviewer: @nedmsmith : We have raised a new issue #199 to address your comments. We can discuss this in the incoming meeting. In the mean time, we intend to apply suggestions from @andrew-draper after clarification from @nedmsmith and acceptance from @thomas-fossati. I hope this is ok, with you.

Please reply with +1 if you agree so that we can un-block it today.

Reviewer 2: +1

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer 5: Given the extensible semantics of the word "matching", I would recommend referencing {{sec-cmp-digests}} here instead of 2 sentences from now for the definition of "matching [...] entry".

For each Integrity Register entry `id => digests_{ref}` in the Reference Value, the Verifier will claim a "match" with Evidence if the Evidence's integrity register entry `id => digests_{ev}` compares positively with the method defined in {{sec-cmp-digests}}. If `id` is not present in the Evidence, then the Reference Value does not "match".

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer 2: ```suggestion
When asserted as Evidence, the Verifier will accept each Integrity Register entry into the ACS.
When asserted as Reference Values, all of the Integrity Register entries of the matching EMT MUST have a corresponding Integrity Register entry in the ACS Evidence.
Additionally, the Verifier will use the associated Integrity Register label (i.e.,`integrity-register-id-type-choice`) to match the label in a corresponding Integrity Register entry in the ACS.
The Verifier compares the labels which MUST exactly match.
If the labels match, then the Integrity Register digest arrays are compared (see {{sec-comid-integrity-registers}} and {{sec-comparison-for-digests-entries}}).
```
Note: {#sec-comparison-for-digets-entries} needs to be added.

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer: @nedmsmith : This text is not clear. We should keep it at a high level describing just the Evidence and Reference Value requirements. Bringing EMT and ACS into mix has made it more complicated, given that we are towards simplification, so this is too much detail without adding much value

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer: I am happy with @andrew-draper suggested `text` here. It is also `sort` of aligned with RATS specification: As mentioned in * Section  8.3 of RFC 9334
`
8.3.  Reference Values

 Reference Values used in appraisal procedures come from a Reference
 Value Provider and are then used by the Verifier to compare to
 Evidence.  Reference Values with matching Evidence produce acceptable
 Claims.  Additionally, an appraisal policy may play a role in
 determining the acceptance of Claims.
`

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer 2: > We should keep it at a high level

We already document the high level rules in earlier sections. It doesn't add value to repeat it here. The reader is potentially uncertain how a verifier will process the array of measurements as each measurement could be processed independently. Alternatively, the collection of measurements could be processed as a singleton measurement. The current text provides no insight or guidance that resolves the reader's confusion.

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer: I quite liked the crisp nature of text @deeglaze added, so tried to modify her text slightly to specify more clarity on 
matching semantics and attempted purely to see how this expresses. See https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/198/files

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer: > > We should keep it at a high level
> 
> We already document the high level rules in earlier sections. It doesn't add value to repeat it here. The reader is potentially uncertain how a verifier will process the array of measurements as each measurement could be processed independently. Alternatively, the collection of measurements could be processed as a singleton measurement. The current text provides no insight or guidance that resolves the reader's confusion.

@nedmsmith : I do not think, we have so far introduced any section as to how `Contents` within Integrity Registers (within Evidence and Reference Values) are to be compared. All I was stating was that: We should not describe the loading of Evidence and  RefValue into ACS specific actions here  and just describe the direct matching semantics of the Registers.

> For each Integrity Register entry in Evidence, the Verifier will use the associated identifier to look up the matching Integrity Register entry in the Reference Value.

Reviewer 2: > We should not describe the loading of Evidence and RefValue into ACS

Line 1869 of my proposed edit could be removed based on this objective. Lines 1870-1873 are aimed at describing the direct matching semantics. Possibly there is better wording, but it seems to say what we want it to say.

If no entry is found, the Reference Value does not match.
Instead, if an entry is found, the digest comparison proceeds as defined in {{sec-cmp-digests}}.

> Note that it is not required for all the entries in the Reference Value to be used during matching: Evidence could consist of a "quote" (in TPM parlance) of just a subset of the register space.

Reviewer 3: ```suggestion
Note that it is not required for all the entries in Evidence to be used during matching: the Reference Value could consist of a "quote" (in TPM parlance) of just a subset of the register space.
```

Reviewer: I am ok, with the suggested text from Ned, as it makes the matching constraints more crisp and consistent.
